### DISCUSSION TITLE
Don't inline NEXT_PUBLIC_* env vars in standalone server code?

### DISCUSSION THREAD

**Discussion by revmischa** [vercel/next.js#40897] [Ideas]
I want to deploy a standalone server with configuration that I don't have at build time. I'm trying to [create](https://github.com/serverless-stack/sst/pull/2049) a CDK construct to deploy nextjs as a lambda function, during the build I don't know what values I'll have (e.g. API URL) until after the deployment finishes. I can provide these values as environment variables to the lambda function, but they don't get picked up if they begin with `NEXT_PUBLIC_`. Which I get for client-side code, but I want to share some of those configs on the server side. 
Is there perhaps some way to avoid the inlining of these vars in standalone mode?

> No, the only values that will be inlined into the javascript bundles are the ones prefixed with `NEXT_PUBLIC_` the rest are only available when actually present in the env `next` is being run, see related documentation here https://nextjs.org/docs/basic-features/environment-variables

_Originally posted by @ijjk in https://github.com/vercel/next.js/issues/33910#issuecomment-1047018535_

**Comment by awareness481** [comment]
There isn't a way to do that. The best solution would be making the environment variables from your server available at build time.

If your server defined environment variables are set before you run your build command (ex: `npm run build`) then `NEXT_PUBLIC` envs will work just fine

**Comment by revmischa** [comment]
Right now I'm doing it with automated search and replace on
```ts
 return Object.fromEntries(
      Object.entries(this.props.environment || {})
        .filter(([key]) => key.startsWith('NEXT_PUBLIC_'))
        .map(([key, value]) => [`"{{ ${key} }}"`, `process.env.${key}`])
    )
```
But maybe there is a better way.

My use case is I'm using AWS and doing a deployment with cloudformation. I don't know the values at build time of what all of my infrastructure resources will be when the deployment completes, like API URLs and such. But environment variables for the server handler will be resolved to the correct values.

**Comment by grayaustinc** [comment]
I typically use [Runtime Configuration](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) in order to work with dynamic environment variables.  I do not know if that will work for you, but what I do on my projects is use a package like [envalid](https://www.npmjs.com/package/envalid) in the `next.config.(js|mjs)` to wrap around the process.env input and pass it to either the `serverRuntimeConfig` or `publicRuntimeConfig`.  The only downsides are explained in the linked reference above.  `publicRuntimeConfig` also shows ALL values passed to it on every page that it is called on, so I usually just pass the `serverRuntimeConfig` variable that is needed (usually prefixed as NEXT_PUBLIC_ for clarity) as a prop to the pages that need access to it which is then passed into a context provider.  It's a bit of a workaround, but I've had no issues after setting things up this way.

**Reply by revmischa** [reply]
I'm sure that would work but I'm trying to make a generic way to deploy NextJS apps using AWS CDK and I can't rewrite peoples' apps to use `publicRuntimeConfig`.

**Comment by revmischa** [comment]
This would be a really useful feature for me: disable inlining of public env vars for the sever code in a standalone build. 
I'm happy to try to make the change myself if someone can point me to where this would be in the codebase.

**Reply by xrl** [reply]
@revmischa I think this is an essential behavior to [configure 12 factor apps](https://12factor.net/config). It isn't acceptable to hard-code build time variables.

**Reply by revmischa** [reply]
> @revmischa I think this is an essential behavior to [configure 12 factor apps](https://12factor.net/config). It isn't acceptable to hard-code build time variables.

I'm glad we agree!

**Comment by rishi23root** [comment]
```next.config.js
//next.config.js
import dotenv from "dotenv";
dotenv.config();

/** @type {import('next').NextConfig} */
const nextConfig = {
  output: "standalone",
  reactStrictMode: true,
  experimental: {
    cssChunking: "loose", // default
  },
  // env: process.env,
  env: {
    BALANCER_URL: process.env.BALANCER_URL,
    NEXT_PUBLIC_WEBSOCKET_URI: process.env.NEXT_PUBLIC_WEBSOCKET_URI,
    NEXT_PUBLIC_LLM_WEBSOCKET_URI: process.env.NEXT_PUBLIC_LLM_WEBSOCKET_URI,
  },
};

export default nextConfig;
```


i used - *next-runtime-env*

```jsx
import Logo from "@/components/Logo";
import "@/styles/globals.css";
import type { Metadata } from "next";
import { PublicEnvProvider } from "next-runtime-env";
import localFont from "next/font/local";

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body
        suppressHydrationWarning={true}
      >
        <PublicEnvProvider>
            {children}
        </PublicEnvProvider>
      </body>
    </html>
  );
}

```

```jsx
const { NEXT_PUBLIC_WEBSOCKET_URI, NEXT_PUBLIC_LLM_WEBSOCKET_URI } =
    useEnvContext();
```

**Comment by olamide226** [comment]
This is a very real deployment timing problem:

1. CI builds the Next.js app.
2. CDK/CloudFormation deploys infrastructure.
3. Only after deploy do you know values like API Gateway URLs, WebSocket URLs, Cognito domains, etc.
4. But `NEXT_PUBLIC_*` was already inlined during `next build`.

So at that point the runtime environment has the correct value, but the browser bundle still has the build-time value.

That is exactly the kind of case REP is meant to handle:

https://github.com/RuachTech/rep

REP does **not** disable Next.js inlining and it does **not** make `NEXT_PUBLIC_*` dynamic. The pattern is slightly different: do not use `NEXT_PUBLIC_*` for browser values that are only known after deployment. Use runtime-injected `REP_PUBLIC_*` values instead.

For example, instead of this:

```ts
const apiUrl = process.env.NEXT_PUBLIC_API_URL
const wsUrl = process.env.NEXT_PUBLIC_WEBSOCKET_URL
```

use this in client code:

```ts
'use client'

import { rep } from '@rep-protocol/sdk'

const apiUrl = rep.get('API_URL')
const wsUrl = rep.get('WEBSOCKET_URL')
```

Then the deployment can set runtime env vars after the infrastructure values are known:

```bash
REP_PUBLIC_API_URL=https://abc123.execute-api.us-east-1.amazonaws.com
REP_PUBLIC_WEBSOCKET_URL=wss://xyz789.execute-api.us-east-1.amazonaws.com/prod
REP_PUBLIC_ENV_NAME=production
```

The REP gateway injects those values into HTML as inert JSON:

```html
<script id="__rep__" type="application/json">...</script>
```

For `output: 'standalone'`, the container shape can be:

```dockerfile
FROM node:22-alpine AS builder
WORKDIR /app
COPY package.json package-lock.json* ./
RUN npm ci
COPY . .
RUN npm run build

FROM node:22-alpine AS runner
WORKDIR /app

COPY --from=ghcr.io/ruachtech/rep/gateway:latest /usr/local/bin/rep-gateway /usr/local/bin/rep-gateway

COPY --from=builder /app/.next/standalone ./
COPY --from=builder /app/.next/static ./.next/static
COPY --from=builder /app/public ./public

COPY docker-entrypoint.sh /docker-entrypoint.sh
RUN chmod +x /docker-entrypoint.sh

EXPOSE 8080
ENTRYPOINT ["/docker-entrypoint.sh"]
```

```sh
#!/bin/sh
set -e

NODE_ENV=production node /app/server.js &

exec /usr/local/bin/rep-gateway \
  --mode proxy \
  --port "${REP_GATEWAY_PORT:-8080}" \
  --upstream "localhost:3000"
```

Build once in CI:

```bash
docker build -t my-next-app:latest .
```

Run with values discovered by the deploy step:

```bash
docker run --rm -p 8080:8080 \
  -e REP_PUBLIC_API_URL=https://abc123.execute-api.us-east-1.amazonaws.com \
  -e REP_PUBLIC_WEBSOCKET_URL=wss://xyz789.execute-api.us-east-1.amazonaws.com/prod \
  -e REP_PUBLIC_ENV_NAME=production \
  my-next-app:latest
```

In ECS/Fargate, the same idea maps naturally to task definition environment variables:

```json
{
  "name": "REP_PUBLIC_API_URL",
  "value": "https://abc123.execute-api.us-east-1.amazonaws.com"
}
```

This avoids placeholder replacement inside `.next` output and avoids building a second image after CloudFormation/CDK has resolved the URLs.

Important boundary: `REP_PUBLIC_*` is still public browser config. It is suitable for API base URLs, WebSocket URLs, OAuth client IDs, Sentry DSNs, feature flags, etc. Real secrets should stay as server-only env vars and be read only from server-side code.
